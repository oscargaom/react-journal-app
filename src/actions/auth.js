import { types } from "../types/types";
import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { startLoading, finishLoading } from "./ui";
import { cleanNotes } from "./notes";

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                const { displayName, uid } = user;
                // console.log(displayName);
                // console.log(uid);
                dispatch(login(displayName, uid));
            });
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        // console.log(email);
        // console.log(password);
        // console.log(name);

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async (userCredential) => {
                // console.log(userCredential);
                const { user } = userCredential;
                await user.updateProfile({ displayName: name });
                const { displayName, uid } = user;
                // console.log(displayName);
                // console.log(uid);

                // console.log(user);
                // console.log(displayName);
                Swal.fire('Bienvenido', 'Tu cuenta ha sido creada', 'success');
                dispatch(login(uid, displayName));

            })
            .catch(error => {
                // console.log(error);
                Swal.fire('Register error', error.message, 'error')
            });
    };
}

/*  Esta es una acción asíncrona, por ello en vez de regresar un objeto regresa 
    un callback el cual el middleware thunk queda en espera de la ejecución de algún
    dispatch; en caso de que no se ejecutará algún dispatch simplmente termina. 
    El dispatch es proporcionado por el middleware thunk, así es que no debemos 
    preocuparnos por obtenerlo de algún otro lugar.
*/
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        // setTimeout(() => {
        //     dispatch(login('Gpe', '456789'));
        // }, 3000);


        dispatch(startLoading());

        // setTimeout(() => {
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then((userCredential) => {
                dispatch(finishLoading());
                const { user: { displayName, uid } } = userCredential;
                Swal.fire('Bienvenido',`Hola ${displayName}`,'success');
                dispatch(login(uid, displayName));
            })
            .catch(error => {
                // const {code} = error;
                dispatch(finishLoading());
                dispatch(login('', ''));
                Swal.fire('Login Error', error.message, 'error');
            });
        // }, 9000);
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout());

        dispatch(cleanNotes());
    };
}



