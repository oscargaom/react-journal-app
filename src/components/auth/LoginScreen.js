import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { useDispatch, useSelector } from 'react-redux';

import { startLoginEmailPassword, startGoogleLogin } from '../../actions/auth';

export const LoginScreen = () => {

    const [formValues, formHandleInputChange] = useForm({
        email: 'oscar@gmail.com',
        password: 'Hola123'
    });

    const { email, password } = formValues;

    const dispatch = useDispatch();

    const { ui: { loading } } = useSelector(state => state);

    // console.log(loading);

    const handleLogin = (e) => {
        e.preventDefault();
        // dispatch(login('987654', 'Manuel'));
        // console.log('dispatch startLoginEmailPassword');
        dispatch(startLoginEmailPassword(email, password));
    };

    const handleGoogleLogin = () => {
        // console.log('dispatch startGoogleLogin');
        dispatch(startGoogleLogin());
    };

    return (
        <>
            <h3 className="auth__title">Login</h3>
            <form 
                onSubmit={handleLogin}
                className="animate__animated animate__pulse animate__faster"
            >
                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={formHandleInputChange}
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={formHandleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block"
                    disabled={loading}
                >
                    Login
                </button>

                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div
                        className="google-btn"
                        onClick={handleGoogleLogin}
                    >
                        <div className="google-icon-wrapper">
                            <img
                                alt="google button"
                                className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg"
                            />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>
                </div>

                <Link
                    to="/auth/register"
                    className="link"
                >
                    Create new account
                </Link>
            </form>
        </>
    )
}
