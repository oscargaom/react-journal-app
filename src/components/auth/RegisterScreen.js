import React from 'react'
import validator from 'validator'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import { setError, removeError } from '../../actions/ui'
import { useDispatch, useSelector } from 'react-redux'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

export const RegisterScreen = () => {

    const [formValues, formHandleInputChange] = useForm({
        name: 'oscar',
        email: 'oscar@gmail.com',
        password: 'Hola123',
        password2: 'Hola123'
    });

    const { name, email, password, password2 } = formValues;

    const dispatch = useDispatch();

    const state = useSelector(state => state);

    const { ui: { msgError } } = state;

    // console.log(msgError);

    const handleRegister = (e) => {

        e.preventDefault();

        if (isFormValid()) {
            // console.log('Formulario correcto');
            dispatch(startRegisterWithEmailPasswordName(email, password, name));
        }
    };

    const isFormValid = () => {

        if (name.trim().length === 0) {
            // console.log('Name is requiered');
            dispatch(setError('Name is requiered'));
            return false;
        } else if (!validator.isEmail(email)) {
            // console.log('Email is required');
            dispatch(setError('Email is required'));
            return false;
        } else if (password !== password2 || password.length < 5) {
            // console.log('Password should be at least 6 characters and match');
            dispatch(setError('Password should be at least 6 characters and match'));
            return false;
        }

        dispatch(removeError());

        return true;
    };

    return (
        <>
            <h3 className="auth__title">Register</h3>
            <form
                onSubmit={handleRegister}
                className="animate__animated animate__pulse animate__faster"
            >

                {
                    msgError !== null && (
                        <div className="auth__alert-error">
                            {msgError}
                        </div>
                    )
                }
                <input
                    type="text"
                    placeholder="Name"
                    name="name"
                    className="auth__input"
                    autoComplete="off"
                    value={name}
                    onChange={formHandleInputChange}
                />

                <input
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={email}
                    onChange={formHandleInputChange}
                />

                <input
                    type="password"
                    placeholder="Password"
                    name="password"
                    className="auth__input"
                    value={password}
                    onChange={formHandleInputChange}
                />

                <input
                    type="password"
                    placeholder="Confirm password"
                    name="password2"
                    className="auth__input"
                    value={password2}
                    onChange={formHandleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                >
                    Register
            </button>


                <Link
                    to="/auth/login"
                    className="link"
                >
                    Already register?
            </Link>
            </form>
        </>
    )
}
