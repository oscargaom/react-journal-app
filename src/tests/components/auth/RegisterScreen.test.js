import React from 'react'
import '@testing-library/jest-dom';
import { mount } from 'enzyme';

import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store'

import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { types } from '../../../types/types';

import { initialStateEmpty, initialStateMsgError } from '../../utils/initialState';
/*  Test debe ejecutarse el dispatch de error al introducir un correo no valido
    haciendo uso de ject.mock()
    
    // import { setError } from '../../../actions/ui';
*/
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

/*  Test debe ejecutarse el dispatch de error al introducir un correo no valido
    haciendo uso de ject.mock()

    // jest.mock('../../../actions/ui', () => ({
    //     setError: jest.fn()
    // }));
*/


describe('Pruebas en el componente <RegisterScreen />', () => {

    let store = mockStore(initialStateEmpty);

    test('debe mostrarse el snashopt correctamente', () => {
        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('debe ejecutarse el dispatch de error al introducir un correo no valido', () => {
        /*  Test debe ejecutarse el dispatch de error al introducir un correo no valido
            haciendo uso de ject.mock()

            // store.dispatch = jest.fn();
        */

        const wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        const emailInput = wrapper.find('input[name="email"]')

        emailInput.simulate('change', {
            target: {
                name: 'email',
                value: ''
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault() { }
        });

        const actions = store.getActions();

        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is required'
        });

        /*  Test debe ejecutarse el dispatch de error al introducir un correo no valido
            haciendo uso de ject.mock()
            // expect(setError).toHaveBeenCalledTimes(1);
            // expect(setError).toHaveBeenCalledWith('Email is required');
        */

    });

    test('debe mostar el mensaje de Sw con el error', () => {

        const localStore = mockStore(initialStateMsgError);

        const wrapper = mount(
            <Provider store={localStore}>
                <MemoryRouter>
                    <RegisterScreen />
                </MemoryRouter>
            </Provider>
        );

        expect(wrapper.find('.auth__alert-error').exists());
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initialStateMsgError.ui.msgError);
    });


});
