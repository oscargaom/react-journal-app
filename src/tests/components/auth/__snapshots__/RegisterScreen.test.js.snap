// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <RegisterScreen /> debe mostrarse el snashopt correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__pulse animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="oscar"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="oscar@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="Hola123"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value="Hola123"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already register?
    </a>
  </form>,
]
`;
